File: Languages/python/algorithms-dsa/leetcode/topological-sort/course_schedule_ii.py
"""
LeetCode #210 - Course Schedule II
Topic: Topological Sort
Difficulty: Medium

Return the ordering of courses you should take to finish all courses.
If impossible (cycle), return empty array.

Approach: Kahn's Algorithm (BFS-based topological sort).
1. Build adjacency list and in-degree array.
2. Start BFS from all nodes with in-degree 0.
3. For each processed node, decrement neighbors' in-degree.
4. If result length != numCourses, there's a cycle.

Time Complexity: O(V + E)
Space Complexity: O(V + E)
"""

from collections import deque


def find_order(num_courses: int, prerequisites: list[list[int]]) -> list[int]:
    graph: list[list[int]] = [[] for _ in range(num_courses)]
    in_degree = [0] * num_courses

    for course, prereq in prerequisites:
        graph[prereq].append(course)
        in_degree[course] += 1

    # Start with courses that have no prerequisites
    queue = deque(i for i in range(num_courses) if in_degree[i] == 0)
    order: list[int] = []

    while queue:
        node = queue.popleft()
        order.append(node)
        for neighbor in graph[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    return order if len(order) == num_courses else []


if __name__ == "__main__":
    assert find_order(2, [[1, 0]]) == [0, 1]
    result = find_order(4, [[1, 0], [2, 0], [3, 1], [3, 2]])
    # Valid topo orders: [0,1,2,3] or [0,2,1,3]
    assert result[0] == 0 and result[-1] == 3
    assert find_order(2, [[1, 0], [0, 1]]) == []  # cycle
    assert find_order(1, []) == [0]
    print("All tests passed!")

File: Languages/cpp/algorithms-dsa/leetcode/topological-sort/course_schedule_ii.cpp
/**
 * LeetCode 210: Course Schedule II
 * Topic: Topological Sort
 * Difficulty: Medium
 *
 * Return the ordering of courses you should take (topological order).
 * Kahn's algorithm: BFS with in-degree tracking.
 * Time: O(V + E), Space: O(V + E)
 */
#include <iostream>
#include <vector>
#include <queue>
#include <cassert>
using namespace std;

vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
    vector<vector<int>> graph(numCourses);
    vector<int> inDegree(numCourses, 0);

    for (auto& p : prerequisites) {
        graph[p[1]].push_back(p[0]);
        inDegree[p[0]]++;
    }

    queue<int> q;
    for (int i = 0; i < numCourses; i++) {
        if (inDegree[i] == 0) q.push(i);
    }

    vector<int> order;
    while (!q.empty()) {
        int course = q.front(); q.pop();
        order.push_back(course);
        for (int next : graph[course]) {
            if (--inDegree[next] == 0) {
                q.push(next);
            }
        }
    }

    return order.size() == (size_t)numCourses ? order : vector<int>{};
}

int main() {
    vector<vector<int>> p1 = {{1,0}};
    auto r1 = findOrder(2, p1);
    assert(r1.size() == 2 && r1[0] == 0 && r1[1] == 1);

    vector<vector<int>> p2 = {{1,0},{2,0},{3,1},{3,2}};
    auto r2 = findOrder(4, p2);
    assert(r2.size() == 4);

    // Cycle: impossible
    vector<vector<int>> p3 = {{1,0},{0,1}};
    auto r3 = findOrder(2, p3);
    assert(r3.empty());

    cout << "All tests passed!" << endl;
    return 0;
}
